android/src/main/java/com/tailscale/ipn/App.kt

```diff
@@ -143,33 +144,36 @@ class App : Application(), libtailscale.AppContext {
   }
 
   // requestNetwork attempts to find the best network that matches the passed NetworkRequest. It is
-  // possible that
-  // this might return an unusuable network, eg a captive portal.
+  // possible that this might return an unusuable network, eg a captive portal.
   private fun setAndRegisterNetworkCallbacks() {
     connectivityManager.requestNetwork(
-        dnsConfigObj.dnsConfigNetworkRequest,
+        networkConnectivityRequest,
         object : ConnectivityManager.NetworkCallback() {
           override fun onAvailable(network: Network) {
             super.onAvailable(network)
+
             val sb = StringBuilder()
             val linkProperties: LinkProperties? = connectivityManager.getLinkProperties(network)
-            val dnsList: MutableList<InetAddress> =
-                linkProperties?.getDnsServers() ?: mutableListOf()
+            val dnsList: MutableList<InetAddress> = linkProperties?.dnsServers ?: mutableListOf()
             for (ip in dnsList) {
-              sb.append(ip.getHostAddress()).append(" ")
+              sb.append(ip.hostAddress).append(" ")
             }
-            val searchDomains: String? = linkProperties?.getDomains()
+            val searchDomains: String? = linkProperties?.domains
             if (searchDomains != null) {
               sb.append("\n")
               sb.append(searchDomains)
             }
-            dnsConfigObj.updateDNSFromNetwork(sb.toString())
-            Libtailscale.onDnsConfigChanged()
+
+            if (dns.updateDNSFromNetwork(sb.toString())) {
```

Comment:
Avoid unnecessary changes, since calling onDnsConfigChanged() triggers a network reconfiguration on the backend.